src/client/clientmap.cpp

```diff
@@ -623,6 +623,14 @@ void ClientMap::updateDrawList()
 		}
 	}
 
+	// Force update meshes of all loaded mapblocks
```

Comment:
This is definitely not needed with shaders enabled (in fact the code that sets these vertex colors only runs if `m_enable_shaders` is `false`).

I also wonder why this is needed at all? You have already hooked `MapBlockMesh::animate`, which should run in time to let you update the vertex colors if shaders are disabled.

Reply:
I've added per-vertex re-calculation of colors in `MapblockMesh::animate()` on demand. However, it feels slower, decreased my FPS before 45 - 50.

Reply:
> However, it feels slower, decreased my FPS before 45 - 50.

Do you mean it decreased your FPS *by* 45 - 50? Or *to* 45 - 50 (*from* what)?

That's certainly odd; that shouldn't happen. One issue that the code still seems to have: It forces an "update" even if shaders are enabled (in which case no update at all is necessary).

Reply:
To 45. And yes, I forgot to add a check for the shaders are enabled in the client env step.